### PR TITLE
[P5-2] Add deterministic clamp_swiglu_weighted CUDA kernel

### PR DESCRIPTION
**Pull request by origamyee0402**
## Summary

Implement P5-2 `clamp_swiglu_weighted` with deterministic CUDA forward and backward kernels.

## Changes

* Added FP32 CUDA forward computation with:

  * `gate` upper clamp at `10`
  * `up` clamp to `[-10, 10]`
  * optional routing weight `p_s`
  * one final BF16 cast
* Added backward kernels for `d_gate`, `d_up`, and deterministic serial `dp_s` reduction.
* Added strict clamp-boundary subgradient behavior.
* Added C++/PyBind bindings in `csrc/ops.cpp`.
* Added `ClampSwiGLUWeightedCudaProvider`.
* Added CUDA tests for boundary cases, shared expert mode, repeated runs, and validation errors.

## Validation

* Black: passed
* Ruff: passed
* CPU provider tests: `3 passed`
* CPU P5 acceptance:

  * `base_only_one_row`: passed
  * `shared_t1`: passed
* CUDA tests: skipped locally because this environment has no NVIDIA GPU.

The golden-manifest anchor test was not used as a local gate because the committed manifest is CPU-x86 specific, while local validation uses Apple Silicon with PyTorch 2.14.0. The manifest was not regenerated or modified.

## Summary by CodeRabbit

- **New Features**
  - Added weighted, clamped SwiGLU activation support with forward and backward operations, including optional route weights and CUDA acceleration.
  - Added a Mixture-of-Experts framework with validated expert data contracts, provider selection, shared and routed expert pipelines, LoRA support, and trace comparison utilities.
  - Added exact MXFP8/MXFP4 quantization and dequantization support.
  - Added deterministic fixtures, golden outputs, and an acceptance-check command for validating provider implementations.

- **Documentation**
  - Added the P5 Expert Start Kit design and numeric compatibility contract.

- **Tests**
  - Added comprehensive coverage for CUDA activation behavior, quantization, contracts, providers, oracle calculations, determinism, validation, and boundary cases.

- **Style**
  - Standardized formatting configuration and reformatted several statements without changing behavior.

### CODE REVIEW

File: csrc/cuda/activation.cu [hunk 1]
```diff
@@ -93,6 +93,164 @@ __global__ void swiglu_backward_kernel(
   d_gate[idx] = static_cast<scalar_t>(dyv * uv * silu_grad_f32(gv));
 }
 
+__device__ __forceinline__ float sigmoid_f32_strict(float x) { // new
+  const float denominator = __fadd_rn(1.0f, expf(-x));
+  return 1.0f / denominator;
+}
```
[the remark this answers is not in the document]
Comment (hsiuyee) [reply csrc/cuda/activation.cu]: The CI linting job fails on examples/vime_qwen3_8b_tp4_cp2_200/collect_results.py, which is not included in this PR diff. The four files changed by this PR pass Black and Ruff locally. Could you please confirm whether this is an existing formatting issue on dsv4-p5-dev?

**Comment by hsiuyee** [comment]
The GitHub linting failure is caused by Black formatting in
examples/vime_qwen3_8b_tp4_cp2_200/collect_results.py, which is not part of
this PR diff. The PR itself changes only four P5-2 files.

CodeRabbit also reports docstring coverage over a broader 21-file review
scope, while GitHub shows only four files changed in this PR. Could you
please confirm whether the lint and docstring checks are evaluating the
correct base branch?

**Comment by hsiuyee** [comment]
@coderabbitai review

**Review by KJLdefeated — COMMENTED** [review]
Verified on an H100 (torch 2.14+cu130), since the CUDA path hadn't been run yet:

- `tests/test_p5_clamp_swiglu_weighted_cuda.py`: 4 passed
- `scripts/check_p5.py --provider rl_engine.moe.cuda_provider:ClampSwiGLUWeightedCudaProvider --device cuda`: PASS, all boundaries byte-equal (routed cases with LoRA and `dp_s` included)

Numerics look right. A few requests before merge:

1. **Saved state is 4x heavier than it needs to be.** Forward writes `g`, `u`, `sig`, `silu` as FP32 `[T, F]`. In shared mode `g`/`u` equal the inputs, and `sig`/`silu` can be recomputed in backward from `gate` alone. Recompute is cheaper than the memory traffic at DSv4 token counts.

2. **Accept the packed `[T, 2F]` layout.** The GEMM feeding this op produces gate|up packed; requiring two contiguous halves forces two copies per call.

3. **Guard against fast-math.** `expf` and `1.0f / denominator` only match `torch.sigmoid` without `--use_fast_math`, which `KERNEL_ALIGN_USE_FAST_MATH=1` enables. Add `#if defined(__USE_FAST_MATH__) #error` next to the kernel.

**Comment by hsiuyee** [comment]
Thanks for the review. I addressed all three requests in the latest commit:

1. Removed `g`, `u`, `sig`, and `silu` from the saved state. Backward now deterministically recomputes them from the saved inputs.
2. Added packed `[T, 2F]` forward and backward CUDA paths while retaining the separate-input path for start-kit compatibility. The new test verifies byte-exact outputs and confirms that a contiguous FP32 packed input reuses the same storage.
3. Added a compile-time `__USE_FAST_MATH__` guard so the strict kernel fails closed when compiled with `--use_fast_math`.

Local formatting, lint, and syntax checks pass. I currently do not have access to the H100 server. Could you please help rebuild the extension from the latest commit and rerun the CUDA test and P5 acceptance command?

**Comment by hsiuyee** [comment]
Update: I regained access to the H100 server and rebuilt the extension from the latest commit (PyTorch 2.14.0+cu130).

* `tests/test_p5_clamp_swiglu_weighted_cuda.py`: **6 passed**
* `scripts/check_p5.py --provider rl_engine.moe.cuda_provider:ClampSwiGLUWeightedCudaProvider --device cuda --cases base_only_one_row,shared_t1`: **PASS**

The packed shared/weighted paths passed, and all checked P5 boundaries remained byte-exact against `oracle-fp32-serial-v1`.

**Review by KJLdefeated — APPROVED** [review]
LGTM
